code/modules/mob/living/carbon/human/species_types/IPC.dm

```diff
@@ -132,53 +132,87 @@
 		return
 
-	if(A.cell && A.cell.charge > A.cell.maxcharge/4)
-		if(H.nutrition >= NUTRITION_LEVEL_ALMOST_FULL)
-			to_chat(user, "<span class='warning'>You are already fully charged!</span>")
+	if(H.nutrition >= NUTRITION_LEVEL_ALMOST_FULL)
+		to_chat(user, "<span class='warning'>You are already fully charged!</span>")
+		return
+
+	if(istype(target, /obj/machinery/power/apc))
+		var/obj/machinery/power/apc/A = target
+		if(A.cell && A.cell.charge > A.cell.maxcharge/4)
+			powerdraw_loop(A, H, TRUE)
 			return
 		else
-			powerdraw_loop(A, H)
+			to_chat(user, "<span class='warning'>There is not enough charge to draw from that APC.</span>")
 			return
 
-	to_chat(user, "<span class='warning'>There is not enough charge to draw from that APC.</span>")
-
-/obj/item/apc_powercord/proc/powerdraw_loop(obj/machinery/power/apc/A, mob/living/carbon/human/H)
-	H.visible_message("<span class='notice'>[H] inserts a power connector into the [A].</span>", "<span class='notice'>You begin to draw power from the [A].</span>")
-	var/obj/item/organ/stomach/battery/battery = locate(/obj/item/organ/stomach/battery) in H.internal_organs
-	while(do_after(H, 10, target = A))
-		if(!battery)
-			to_chat(H, "<span class='warning'>You need a battery to recharge!</span>")
-			break
-		if(loc != H)
-			to_chat(H, "<span class='warning'>You must keep your connector out while charging!</span>")
-			break
-		if(A.cell.charge <= A.cell.maxcharge/4)
-			to_chat(H, "<span class='warning'>The [A] doesn't have enough charge to spare.</span>")
-			break
-		A.charging = 1
-		if(A.cell.charge > A.cell.maxcharge/4 + 250)
-			battery.adjust_charge(250)
-			A.cell.charge -= 250
-			to_chat(H, "<span class='notice'>You siphon off some of the stored charge for your own use.</span>")
+	if(isethereal(target))
+		var/mob/living/carbon/human/target_ethereal = target
+		var/obj/item/organ/stomach/battery/target_battery = target_ethereal.getorganslot(ORGAN_SLOT_STOMACH)
+		if(target_ethereal.nutrition > 0)
+			powerdraw_loop(target_battery, H, FALSE)
+			return
 		else
-			battery.adjust_charge(A.cell.charge - A.cell.maxcharge/4)
-			A.cell.charge = A.cell.maxcharge/4
-			to_chat(H, "<span class='notice'>You siphon off as much as the [A] can spare.</span>")
-			break
-		if(battery.charge >= battery.max_charge)
-			to_chat(H, "<span class='notice'>You are now fully charged.</span>")
-			break
-	H.visible_message("<span class='notice'>[H] unplugs from the [A].</span>", "<span class='notice'>You unplug from the [A].</span>")
+			to_chat(user, "<span class='warning'>There is not enough charge to draw from that being!</span>")
+			return
+/obj/item/apc_powercord/proc/powerdraw_loop(atom/target, mob/living/carbon/human/H, apc_target)
+	H.visible_message("<span class='notice'>[H] inserts a power connector into [target].</span>", "<span class='notice'>You begin to draw power from the [target].</span>")
```

Comment:
if you're drawing from an ethereal, will this say that you're drawing power from "the Aurora IV" or the like?

Reply:
Please read the code. Powerdraw loop is passed either an APC or an ethereal stomach, not an ethereal itself.